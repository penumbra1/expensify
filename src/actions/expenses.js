import database from "../firebase/firebase";

export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

// Instead of an action object, startAddExpense returns an "action promise"
// that takes a dispatch argument and will be processed by Thunk middleware
export const startAddExpense = (expenseData = {}) => dispatch => {
  const {
    description = "",
    note = "",
    amount = 0,
    createdAt = 0
  } = expenseData;
  const expense = { description, note, amount, createdAt };
  return database
    .ref("expenses")
    .push(expense)
    .then(data => dispatch(addExpense({ id: data.key, ...expense })))
    .catch(e => console.log("Failed to add expense", e));
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id }) => dispatch => {
  database
    .ref(`expenses/${id}`)
    .remove()
    .then(() => dispatch(removeExpense({ id })))
    .catch(e => console.log("Failed to remove expense", e));
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = ({ id, ...updates }) => dispatch =>
  database
    .ref(`expenses/${id}`)
    .update(updates)
    .then(() => dispatch(editExpense(id, updates)))
    .catch(e => console.log("Failed to update expense", e));
