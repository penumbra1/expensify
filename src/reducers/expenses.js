import { combineReducers } from "redux";

const byIdReducer = (state = {}, action) => {
  switch (action.type) {
    case "LOAD_EXPENSES":
      // Rewrite any previous state
      return action.expenses;
    case "ADD_EXPENSE":
    case "EDIT_EXPENSE":
      return { ...state, [action.id]: action.data };
    case "REMOVE_EXPENSE": {
      const { [action.id]: removed, ...remaining } = state;
      return remaining;
    }
    default:
      return state;
  }
};

const allIdsReducer = (state = [], action) => {
  switch (action.type) {
    case "LOAD_EXPENSES":
      return Object.keys(action.expenses);
    case "ADD_EXPENSE":
      return [...state, action.id];
    case "REMOVE_EXPENSE":
      return state.filter(id => id !== action.id);
    default:
      return state;
  }
};

export default combineReducers({ byId: byIdReducer, allIds: allIdsReducer });
