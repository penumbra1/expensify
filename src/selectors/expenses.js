import moment from "moment";

export default (expenses, { text, sortBy, startDate, endDate }) => {
  const sortFunction = (a, b) => {
    switch (sortBy) {
      case "amount":
        return b.amount - a.amount;
      case "date":
        return b.createdAt - a.createdAt;
      // The store can only set sortBy to date or amount
      // Any other case would return undefined, and sort(undefined) is alphabetic
      default:
        return undefined;
    }
  };

  const filterFunction = expense => {
    const createdAtMoment = moment(expense.createdAt);
    // User can clear dates, and no date means a match for all items
    const startDateMatch = startDate
      ? startDate.isSameOrBefore(createdAtMoment, "day")
      : true;
    const endDateMatch = endDate
      ? endDate.isSameOrAfter(createdAtMoment, "day")
      : true;
    const textMatch = text
      ? expense.description.toLowerCase().includes(text.toLowerCase())
      : true;
    return startDateMatch && endDateMatch && textMatch;
  };

  return expenses.filter(filterFunction).sort(sortFunction);
};
