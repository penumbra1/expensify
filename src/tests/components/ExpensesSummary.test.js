import React from "react";
import { shallow } from "enzyme";
import {
  ExpensesSummary,
  mapStateToProps
} from "../../components/ExpensesSummary";
import getExpenseTotal from "../../selectors/expenses-total";
import expenses from "../fixtures/expenses";
import { filtersWithData } from "../fixtures/filters";

test("should render ExpensesSummary correctly with no data", () => {
  const wrapper = shallow(<ExpensesSummary />);

  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary correctly with a single expense", () => {
  const wrapper = shallow(
    <ExpensesSummary expenseCount={1} expenseTotal={expenses[0].amount} />
  );

  expect(
    wrapper
      .find("h2")
      .text()
      .split(" ")[2]
  ).toBe("expense");
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary correctly with multiple expenses", () => {
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={expenses.length}
      expenseTotal={getExpenseTotal(expenses)}
    />
  );
  expect(
    wrapper
      .find("h2")
      .text()
      .split(" ")[2]
  ).toBe("expenses");
  expect(wrapper).toMatchSnapshot();
});

test("should map state to props correctly", () => {
  const newProps = mapStateToProps({ expenses, filters: filtersWithData });
  const wrapper = shallow(<ExpensesSummary {...newProps} />);
  expect(wrapper).toMatchSnapshot();
});
