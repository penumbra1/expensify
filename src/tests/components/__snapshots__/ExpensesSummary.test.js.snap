// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should map state to props correctly 1`] = `
<h2>
  Viewing 1 expense totalling $45.00
</h2>
`;

exports[`should render ExpensesSummary correctly with a single expense 1`] = `
<h2>
  Viewing 1 expense totalling $0.01
</h2>
`;

exports[`should render ExpensesSummary correctly with multiple expenses 1`] = `
<h2>
  Viewing 3 expenses totalling $10,140.01
</h2>
`;

exports[`should render ExpensesSummary correctly with no data 1`] = `""`;
